src/jit/lower.cpp

```diff
@@ -1990,6 +1990,7 @@ GenTree* Lowering::LowerTailCallViaHelper(GenTreeCall* call, GenTree* callTarget
 //
 // Notes:
 //    - Decomposes long comparisons that feed a GT_JTRUE (32 bit specific).
+//    - Decomposes long comparisons that produce a value (X86 specific).
```

Comment:
> There is precedent for checking the user of a freshly-decomposed node and then decomposing the user as well; this is how we handle GT_LONG nodes that are part of field list arguments.

That sounds like a special case and not something that decomposition usually does.

I'd love to actually move all this code to decomposition but I don't see a natural way of doing it, short of simply extracting the code into a separate function and moving the function to `DecomposeLongs`.

> rather than today's

I think that we actually used the first version in the early days of decomposition. I switched to the second version to fix a bug. If memory serves me well the bug was related to embedded (shudder) trees so the current approach version may be obsolete.